src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -375,6 +367,21 @@ public static NoSuchElementException newVaListNSEE(MemoryLayout layout) {
         return new NoSuchElementException("No such element: " + layout);
     }
 
+    // Current limitation of the implementation:
+    // We don't support packed structs on some platforms,
+    // so reject them here explicitly
+    public static void checkHasNaturalAlignment(MemoryLayout layout) {
+        long naturalByteAlignmnet = naturalByteAlignment(layout);
+        if (naturalByteAlignmnet != layout.byteAlignment()) {
+            throw new IllegalArgumentException("Layout bit alignment must be natural alignment: "
+                    + layout.byteAlignment() + " != " + naturalByteAlignmnet);
+        }
+    }
+
+    private static long naturalByteAlignment(MemoryLayout layout) {
+        return alignment(layout, true);
```

Comment:
Can you please take a look at that `alignment` routine? I'm not 100% sure it's correct (e.g. for arrays it returns 16 - is that true on all platforms?). Also, note that `AbstractLayout` has an `hasNaturalAlignment` predicate, which seems close to what you need.

Reply:
Hmm yeah, IIRC the array alignment is only 16 if it is a global variable, so I think we can drop that here in this case. I'll use `hasNaturalAlignment` instead. Good suggestion!